avoidance/src/avoidance_node.cpp

```diff
@@ -70,4 +82,82 @@ void AvoidanceNode::checkFailsafe(ros::Duration since_last_cloud, ros::Duration
     }
   }
 }
-}
\ No newline at end of file
+
+void AvoidanceNode::px4ParamsCallback(const mavros_msgs::Param& msg) {
+  // collect all px4_ parameters needed for model based trajectory planning
+  // when adding new parameter to the struct ModelParameters,
+  // add new else if case with correct value type
+  auto parse_param_f = [&msg](const std::string& name, float& val) -> bool {
+    if (msg.param_id == name) {
+      ROS_INFO("parameter %s is set from  %f to %f \n", name.c_str(), val, msg.value.real);
+      val = msg.value.real;
+      return true;
+    }
+    return false;
+  };
+
+  auto parse_param_i = [&msg](const std::string& name, int& val) -> bool {
+    if (msg.param_id == name) {
+      ROS_INFO("parameter %s is set from %i to %li \n", name.c_str(), val, msg.value.integer);
+      val = msg.value.integer;
+      return true;
+    }
+    return false;
+  };
+
+  // clang-format off
+  parse_param_f("MPC_ACC_DOWN_MAX", px4_.param_mpc_acc_down_max) ||
+  parse_param_f("MPC_ACC_HOR", px4_.param_mpc_acc_hor) ||
+  parse_param_f("MPC_ACC_UP_MAX", px4_.param_acc_up_max) ||
+  parse_param_i("MPC_AUTO_MODE", px4_.param_mpc_auto_mode) ||
+  parse_param_f("MPC_JERK_MIN", px4_.param_mpc_jerk_min) ||
+  parse_param_f("MPC_JERK_MAX", px4_.param_mpc_jerk_max) ||
+  parse_param_f("MPC_LAND_SPEED", px4_.param_mpc_land_speed) ||
+  parse_param_f("MPC_TKO_SPEED", px4_.param_mpc_tko_speed) ||
+  parse_param_f("MPC_XY_CRUISE", px4_.param_mpc_xy_cruise) ||
+  parse_param_f("MPC_Z_VEL_MAX_DN", px4_.param_mpc_vel_max_dn) ||
+  parse_param_f("MPC_Z_VEL_MAX_UP", px4_.param_mpc_z_vel_max_up) ||
+  parse_param_f("MPC_COL_PREV_D", px4_.param_mpc_col_prev_d) ||
+  parse_param_f("NAV_ACC_RAD", px4_.param_nav_acc_rad);
+  // clang-format on
+}
+
+void AvoidanceNode::checkPx4Parameters() {
+  auto& client = get_px4_param_client_;
+  auto request_param = [&client](const std::string& name, float& val) {
+    mavros_msgs::ParamGet req;
+    req.request.param_id = name;
+    if (client.call(req) && req.response.success) {
+      val = req.response.value.real;
+    }
+  };
+  while (!should_exit_) {
+    request_param("MPC_XY_CRUISE", px4_.param_mpc_xy_cruise);
```

Comment:
Based on the MAVROS implementation, this doesn't really make sense. We should probably keep this around for if/when we improve the MAVROS implementation though.

Reply:
I think it works once the entire list of parameters is retrieved once. See planner output on the callbacks 
``` [ERROR] [1562788110.832861864, 9.380000000]: PR: Unknown parameter to get: MPC_XY_CRUISE
[ERROR] [1562788110.834822109, 9.384000000]: PR: Unknown parameter to get: MPC_COL_PREV_D
[ERROR] [1562788110.836733009, 9.384000000]: PR: Unknown parameter to get: MPC_LAND_SPEED
[ERROR] [1562788110.838620490, 9.388000000]: PR: Unknown parameter to get: NAV_ACC_RAD
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
MAVLINK_MSG_ID_PARAM_REQUEST_LIST
px4ParamsCallback 
px4_.param_mpc_xy_cruise nan
px4ParamsCallback 
px4ParamsCallback 
px4_.param_mpc_xy_cruise nan
px4ParamsCallback 
px4ParamsCallback 
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise nan
request single param 
px4_.param_mpc_xy_cruise nan
px4_.param_mpc_xy_cruise 5.000000
px4_.param_mpc_xy_cruise 5.000000
 ```

Reply:
Yes, what I mean is that MAVROS doesn't re-request the info from the FCU once we ask for these params. It just serves it from its local cache which it tries to keep up to date with the FCU just via the general subscription method, not polling the exact parameters we ask for.